### ISSUE TITLE
Definition of CYTHON_CCOMPLEX

### ISSUE THREAD

**Issue by realead** [cython/cython#2513]
CCOMPLEX_CYTHON is defined [as follows](https://github.com/cython/cython/blob/e6315294395ad44e667a804710938ee799575296/Cython/Utility/Complex.c#L4):

```
#if !defined(CYTHON_CCOMPLEX)
  #if defined(__cplusplus)
    #define CYTHON_CCOMPLEX 1
  #elif defined(_Complex_I)
    #define CYTHON_CCOMPLEX 1
  #else
    #define CYTHON_CCOMPLEX 0
  #endif
#endif

#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    #include <complex>
  #else
    #include <complex.h>
  #endif
#endif
```

If I understand it correctly `_Complex_I` is defined in `complex.h`, so basically for C `CYTHON_COMPLEX` is always 0 (as long as `Python.h` doesn't include complex.h, which seems not to be the case at least with Python-3.6) and thus the resulting code doesn't uses `double complex` in this case but the [Cython's fall back solution](https://github.com/cython/cython/blob/e6315294395ad44e667a804710938ee799575296/Cython/Utility/Complex.c#L63).

`complex.h` is here since C99, so maybe something like this would make sense?

```
#ifdef __cplusplus
    #include <complex>
#elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
     #include <complex.h>
#endif

#if !defined(CYTHON_CCOMPLEX)
  #if defined(__cplusplus)
    #define CYTHON_CCOMPLEX 1
  #elif defined(_Complex_I)
    #define CYTHON_CCOMPLEX 1
  #else
    #define CYTHON_CCOMPLEX 0
  #endif
#endif
```

**Comment by scoder** [comment]
My _guess_ is that the original intention was something like: "if the user includes 'complex.h' then use it automatically".

I don't have a problem with using a header that is guaranteed to be there – as long as it does not get in the way of something that users might want instead.

In any case, though, `CYTHON_CCOMPLEX` should always be checked first, since users can set it to 1/0 if they feel like it.

**Comment by realead** [comment]
As always, this doesn't seem to be clearcut: The way (at least) the multiplication of complex numbers is defined in C99 is not the way this multiplication is implemented in CPython3. 

I think the way of least surprise would be to use the Cython's fallback per default (for C and CPP backends) and only switch to native complex operations is the user explicitly triggers it.

For exaple, the result of the following multiplication

`(1.0+0.0j)*(inf+inf*j)`

is:

**Python**:

   Python uses the formula from the school and gets: `nan+nan*j` as result.

**On Linux:**

  1. C-backend, gcc and clang are C99 compliant. As result, for example `inf+inf*j` is  possible.
  2. CPP-backend. As far as I know (don't take my word for it), c++-standard doesn't define the complex multiplication. gcc and clang use C99-variant. A possible result is like above `inf+inf*j`

Thus on Linux + Cython we have the following behavior:

```
%%cython
def c_mult(double complex a, double complex b):
    return a*b

%%cython --cplus
def cpp_mult(double complex a, double complex b):
    return a*b

a=1.0+0.0j
b=(1.0+1.0j)*float("inf")
print("Python", a*b)
print("C-Cython", c_mult(a,b))
print("Cpp-Cython", cpp_mult(a,b))
```
```
Python (nan+nanj)
C-Cython (nan+nanj)
Cpp-Cython (inf+infj)
```

C-version has the same behavior as CPython, because it uses Cython-fallback. CPP-version uses the native multiplication and the result isn't consistent with other two.

** On Windows **:
 
1. C-backend. MSVC doesn't support complex from C99 (At least with 2015 version), so Cython has to use its fallback here in any case.
2. CPP-backend. The c++-complex multiplication of MSVC doesn't  comply with C99 (via cpp-standard it doesn't have to). It uses the simple formula from the school as far as I understand.

Thus, C-backend and CPP-backend are consistent with CPython implementation on Windows.

**Comment by robertwb** [comment]
Yes, the intent is by default to have complex arithmetic support even if `complex.h` isn't available (e.g. MSVC for C) but use the "native" complex if `complex.h` is imported, and furthermore allow an end user to explicitly choose by defining `CYTHON_CCOMPLEX`.

That's a good point about arithmetic for degenerate values being handled differently, but it seems we already have inconsistency here between the various non-cython implementations already so it's unclear what we must agree with (vs. leaving it undefined) and should probably be a separate discussion/bug.

**Comment by leofang** [comment]
possibly related: #2891

**Comment by da-woods** [comment]
I've gone for a slightly modified version of this. C11 adds a specific macro to check for complex.h (`__STDC_NO_COMPLEX__`) and makes it optional for a "freestanding" implementation (probably an embedded system). Therefore I've turned it on given C11, if the `__STDC_NO_COMPLEX__` macro isn't defined, and retained the option for the user to turn it on/off themselves.

I think that's a fairly safe option that covers the "use it by default if available but let the user override it" intent.

**Comment by scoder** [comment]
Closed by #5021
